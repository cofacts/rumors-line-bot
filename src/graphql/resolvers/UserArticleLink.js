export default {
  /**
   * @param {UserArticleLink} source
   */
  article(source) {
    // Returns stub
    return { id: source.articleId };
  },
};
