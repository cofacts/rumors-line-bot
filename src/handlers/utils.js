export function createPostbackAction(label, input, issuedAt) {
  return {
    type: 'postback',
    label,
    data: JSON.stringify({
      input,
      issuedAt,
    }),
  };
}

export function createFeedbackWords(feedbacks) {
  let positive = 0;
  let negative = 0;
  feedbacks.forEach(e => {
    if (e.score > 0) {
      positive++;
    }
    if (e.score < 0) {
      negative++;
    }
  });
  if (positive + negative === 0) return '[還沒有人針對此回應評價]';
  let result = '';
  if (positive) result += `有 ${positive} 人覺得此回應有幫助\n`;
  if (negative) result += `有 ${negative} 人覺得此回應沒幫助\n`;
  return `[${result.trim()}]`;
}

export function createTypeWords(type) {
  switch (type) {
    case 'RUMOR':
      return '❌ 含有不實訊息';
    case 'NOT_RUMOR':
      return '⭕ 含有真實訊息';
    case 'OPINIONATED':
      return '💬 含有個人意見';
    case 'NOT_ARTICLE':
      return '⚠️️ 不在查證範圍';
  }
  return '回應的狀態未定義！';
}
export function createReferenceWords(reference) {
  if (reference) return `出處：${reference}`;
  return '\uDBC0\uDC85 ⚠️️ 此回應沒有出處，請自行斟酌回應真實。⚠️️  \uDBC0\uDC85';
}
