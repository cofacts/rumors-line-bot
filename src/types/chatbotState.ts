import type { Message, MessageEvent } from '@line/bot-sdk';

export type ChatbotState =
  | '__INIT__'
  | 'TUTORIAL'
  | 'CHOOSING_ARTICLE'
  | 'CHOOSING_REPLY'
  | 'ASKING_ARTICLE_SOURCE'
  | 'ASKING_ARTICLE_SUBMISSION_CONSENT'
  | 'Error';

export type Context = {
  /** Used to differientiate different search sessions (searched text or media) */
  sessionId: number;
} & (
  | {
      /** Searched multi-media message that started this search session */
      messageId: MessageEvent['message']['id'];
      messageType: MessageEvent['message']['type'];
    }
  | {
      /** Searched text that started this search session */
      searchedText: string;
    }
);

/** A single messages in the same co-occurrence */
export type CooccurredMessage = {
  id: MessageEvent['message']['id'];
} & (
  | {
      type: Extract<
        MessageEvent['message']['type'],
        'audio' | 'video' | 'image'
      >;
    }
  | {
      type: Extract<MessageEvent['message']['type'], 'text'>;
      /** Searched text that started this search session */
      text: string;
    }
);

export type ChatbotStateHandlerReturnType = {
  data: Context;
  replies: Message[];
};

/**
 * The data that postback action stores as JSON.
 */
export type PostbackActionData<T> = {
  input: T;
  sessionId: number;
  state: ChatbotState;
};

export type ChatbotPostbackHandlerParams<T = unknown> = {
  /** Data stored in Chatbot context */
  data: Context;
  /** Data in postback payload */
  postbackData: PostbackActionData<T>;
  userId: string;
};

/**
 * For chatbot postback event handers
 */
export type ChatbotPostbackHandler<T = unknown> = (
  params: ChatbotPostbackHandlerParams<T>
) => Promise<ChatbotStateHandlerReturnType>;
