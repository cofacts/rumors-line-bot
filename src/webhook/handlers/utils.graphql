fragment CreateReferenceWordsReply on Reply {
  reference
  type
  createdAt
}

fragment CreateReplyMessagesReply on Reply {
  text
  ...CreateReferenceWordsReply
}

fragment CreateReplyMessagesArticle on Article {
  replyCount
  createdAt
}

fragment CreateHighlightContentsHighlight on Highlights {
  text
  hyperlinks {
    title
    summary
  }
}