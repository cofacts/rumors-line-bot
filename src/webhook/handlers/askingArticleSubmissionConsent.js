import { t } from 'ttag';
import ga from 'src/lib/ga';
import gql from 'src/lib/gql';
import { SOURCE_PREFIX, getArticleURL } from 'src/lib/sharedUtils';
import {
  ManipulationError,
  createArticleShareBubble,
  createSuggestOtherFactCheckerReply,
  getArticleSourceOptionFromLabel,
  createReasonButtonFooter,
} from './utils';

import UserArticleLink from '../../database/models/userArticleLink';

export default async function askingArticleSubmissionConsent(params) {
  let { data, state, event, userId, replies, isSkipUser } = params;

  if (!event.input.startsWith(SOURCE_PREFIX)) {
    throw new ManipulationError(
      t`Please press the latest button to submit message to database.`
    );
  }

  const visitor = ga(userId, state, data.searchedText);

  const sourceOption = getArticleSourceOptionFromLabel(
    event.input.slice(SOURCE_PREFIX.length)
  );

  visitor.event({
    ec: 'UserInput',
    ea: 'ProvidingSource',
    el: sourceOption.value,
  });

  if (!sourceOption.valid) {
    replies = [
      {
        type: 'text',
        text: t`Thanks for the info.`,
      },
      createSuggestOtherFactCheckerReply(),
    ];
    state = '__INIT__';
  } else {
    visitor.event({ ec: 'Article', ea: 'Create', el: 'Yes' });

    const {
      data: { CreateArticle },
    } = await gql`
      mutation($text: String!) {
        CreateArticle(text: $text, reference: { type: LINE }) {
          id
        }
      }
    `({ text: data.searchedText }, { userId });

    await UserArticleLink.create({ userId, articleId: CreateArticle.id });

    const articleUrl = getArticleURL(CreateArticle.id);
    const articleCreatedMsg = t`Your submission is now recorded at ${articleUrl}`;

    // Track the source of the new message.
    visitor.event({
      ec: 'Article',
      ea: 'ProvidingSource',
      el: `${CreateArticle.id}/${sourceOption.value}`,
    });

    replies = [
      {
        type: 'flex',
        altText: articleCreatedMsg,
        contents: {
          type: 'carousel',
          contents: [
            {
              type: 'bubble',
              body: {
                type: 'box',
                layout: 'vertical',
                contents: [
                  {
                    type: 'text',
                    wrap: true,
                    text: articleCreatedMsg,
                  },
                ],
              },
              footer: createReasonButtonFooter(
                articleUrl,
                userId,
                data.sessionId
              ),
            },
            createArticleShareBubble(articleUrl),
          ],
        },
      },
    ];

    // Record article ID in context for reason LIFF
    data.selectedArticleId = CreateArticle.id;
    state = '__INIT__';
  }

  visitor.send();
  return { data, state, event, userId, replies, isSkipUser };
}
