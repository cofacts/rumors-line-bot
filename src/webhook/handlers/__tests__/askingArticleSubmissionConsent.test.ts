jest.mock('src/lib/gql');
jest.mock('src/lib/ga');
import type { MockedGa } from 'src/lib/__mocks__/ga';
import type { MockedGql } from 'src/lib/__mocks__/gql';

import MockDate from 'mockdate';
import askingArticleSubmissionConsent from '../askingArticleSubmissionConsent';
import { POSTBACK_NO, POSTBACK_YES } from '../utils';
import originalGql from 'src/lib/gql';
import originalGa from 'src/lib/ga';

const ga = originalGa as MockedGa;
const gql = originalGql as MockedGql;

import UserSettings from 'src/database/models/userSettings';
import UserArticleLink from 'src/database/models/userArticleLink';
import { ChatbotPostbackHandlerParams } from 'src/types/chatbotState';

beforeAll(async () => {
  if (await UserArticleLink.collectionExists()) {
    await (await UserArticleLink.client).drop();
  }
});

beforeEach(() => {
  ga.clearAllMocks();
});

it('throws on incorrect input', async () => {
  const incorrectParam: ChatbotPostbackHandlerParams = {
    data: { sessionId: 0, searchedText: 'foo' },
    postbackData: {
      sessionId: 0,
      state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
      input: 'Wrong',
    },
    userId: 'the-user-id',
  };

  expect(
    askingArticleSubmissionConsent(incorrectParam)
  ).rejects.toMatchInlineSnapshot(
    `[Error: Please choose from provided options.]`
  );
});

it('should thank the user if user does not agree to submit', async () => {
  const inputSession = new Date('2020-01-01T18:10:18.314Z').getTime();
  const params: ChatbotPostbackHandlerParams = {
    data: {
      sessionId: inputSession,
      searchedText: 'Some text forwarded by the user',
    },
    postbackData: {
      sessionId: inputSession,
      state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
      input: POSTBACK_NO,
    },
    userId: 'userId',
  };

  const { replies } = await askingArticleSubmissionConsent(params);
  expect(replies).toMatchInlineSnapshot(`
    Array [
      Object {
        "altText": "The message has not been reported and won’t be fact-checked. Thanks anyway!",
        "contents": Object {
          "body": Object {
            "contents": Array [
              Object {
                "text": "The message has not been reported and won’t be fact-checked. Thanks anyway!",
                "type": "text",
                "wrap": true,
              },
            ],
            "layout": "vertical",
            "type": "box",
          },
          "type": "bubble",
        },
        "type": "flex",
      },
    ]
  `);
});

it('should submit article if user agrees to submit', async () => {
  const inputSession = new Date('2020-01-01T18:10:18.314Z').getTime();
  const params: ChatbotPostbackHandlerParams = {
    data: {
      sessionId: inputSession,
      searchedText: 'Some text forwarded by the user',
    },
    postbackData: {
      sessionId: inputSession,
      state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
      input: POSTBACK_YES,
    },
    userId: 'userId',
  };

  MockDate.set('2020-01-02');
  gql.__push({ data: { CreateArticle: { id: 'new-article-id' } } });
  // The case when have AI replies
  gql.__push({ data: { CreateAIReply: { text: 'Hello from ChatGPT' } } });
  const result = await askingArticleSubmissionConsent(params);
  MockDate.reset();
  expect(gql.__finished()).toBe(true);

  expect(result).toMatchSnapshot('has AI reply');
  expect(result.data.sessionId).not.toEqual(inputSession);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "Create",
          "ec": "Article",
          "el": "Yes",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);

  // The case when no AI reply is provided (such as in the case of insufficient data)
  //
  MockDate.set('2020-01-02');
  gql.__push({ data: { CreateArticle: { id: 'new-article-id-2' } } });
  gql.__push({ data: { CreateAIReply: null } });
  expect(await askingArticleSubmissionConsent(params)).toMatchSnapshot(
    'has no AI reply'
  );
  MockDate.reset();
  expect(gql.__finished()).toBe(true);
});

it('should submit image article if user agrees to submit', async () => {
  const inputSession = new Date('2020-01-01T18:10:18.314Z').getTime();
  const params: ChatbotPostbackHandlerParams = {
    data: {
      sessionId: inputSession,
      searchedText: '',
      messageId: '6530038889933',
      messageType: 'image',
    },
    postbackData: {
      sessionId: inputSession,
      state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
      input: POSTBACK_YES,
    },
    userId: 'userId',
  };

  MockDate.set('2020-01-02');
  gql.__push({ data: { CreateMediaArticle: { id: 'new-article-id' } } });
  const result = await askingArticleSubmissionConsent(params);
  MockDate.reset();
  expect(gql.__finished()).toBe(true);

  expect(result).toMatchSnapshot();
  expect(result.data.sessionId).not.toEqual(inputSession);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "Create",
          "ec": "Article",
          "el": "Yes",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('should create a UserArticleLink when creating a Article', async () => {
  const userId = 'user-id-0';
  const params: ChatbotPostbackHandlerParams = {
    data: {
      sessionId: 0,
      searchedText: 'Some text forwarded by the user',
    },
    postbackData: {
      sessionId: 0,
      input: POSTBACK_YES,
      state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
    },
    userId,
  };

  MockDate.set('2020-01-01');
  gql.__push({ data: { CreateArticle: { id: 'new-article-id' } } });
  gql.__push({ data: { CreateAIReply: { text: 'Hello from ChatGPT' } } });
  await askingArticleSubmissionConsent(params);
  MockDate.reset();

  const userArticleLinks = await UserArticleLink.findByUserId(userId);
  expect(userArticleLinks.map((e) => ({ ...e, _id: '_id' }))).toMatchSnapshot();
});

it('should ask user to turn on notification settings if they did not turn it on after creating an Article', async () => {
  const userId = 'user-id-0';
  const params: ChatbotPostbackHandlerParams = {
    data: {
      sessionId: 0,
      searchedText: 'Some text forwarded by the user',
    },
    postbackData: {
      sessionId: 0,
      state: 'ASKING_ARTICLE_SUBMISSION_CONSENT',
      input: POSTBACK_YES,
    },
    userId,
  };

  gql.__push({ data: { CreateArticle: { id: 'new-article-id' } } });
  gql.__push({ data: { CreateAIReply: { text: 'Hello from ChatGPT' } } });
  process.env.NOTIFY_METHOD = 'LINE_NOTIFY';
  await UserSettings.setAllowNewReplyUpdate(userId, false);

  MockDate.set('2020-01-01');
  const results = await askingArticleSubmissionConsent(params);
  MockDate.reset();

  const lastReply = results.replies[4];

  // Make TS happy
  /* istanbul ignore if */
  if (lastReply.type !== 'flex' || lastReply.contents.type !== 'carousel') {
    throw new Error('Wrong type for last reply');
  }

  expect(lastReply.contents.contents).toMatchSnapshot();

  delete process.env.NOTIFY_METHOD;
  await UserSettings.setAllowNewReplyUpdate(userId, true);
});
