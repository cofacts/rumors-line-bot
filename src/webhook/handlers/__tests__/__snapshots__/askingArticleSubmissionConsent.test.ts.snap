// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should ask user to turn on notification settings if they did not turn it on after creating an Article 1`] = `
Array [
  Object {
    "body": Object {
      "contents": Array [
        Object {
          "text": "Is the AI analysis helpful?",
          "type": "text",
          "wrap": true,
        },
      ],
      "layout": "vertical",
      "paddingAll": "xl",
      "type": "box",
    },
    "footer": Object {
      "contents": Array [
        Object {
          "action": Object {
            "label": "üëç Yes",
            "type": "uri",
            "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=aiReplyFeedback&aiResponseId=undefined&vote=UPVOTE",
          },
          "color": "#00B172",
          "style": "primary",
          "type": "button",
        },
        Object {
          "action": Object {
            "label": "üòï No",
            "type": "uri",
            "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=aiReplyFeedback&aiResponseId=undefined&vote=DOWNVOTE",
          },
          "color": "#FB5959",
          "style": "primary",
          "type": "button",
        },
      ],
      "layout": "vertical",
      "spacing": "sm",
      "type": "box",
    },
    "header": Object {
      "contents": Array [
        Object {
          "color": "#00B172",
          "size": "lg",
          "text": "Provide feedback to AI analysis",
          "type": "text",
        },
      ],
      "layout": "vertical",
      "paddingBottom": "none",
      "type": "box",
    },
    "type": "bubble",
  },
  Object {
    "body": Object {
      "contents": Array [
        Object {
          "text": "It would help fact checkers a lot if you provide more detail :)",
          "type": "text",
          "wrap": true,
        },
      ],
      "layout": "vertical",
      "type": "box",
    },
    "footer": Object {
      "contents": Array [
        Object {
          "action": Object {
            "label": "Provide detail",
            "type": "uri",
            "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=comment&articleId=new-article-id",
          },
          "color": "#00B172",
          "style": "primary",
          "type": "button",
        },
      ],
      "layout": "vertical",
      "spacing": "sm",
      "type": "box",
    },
    "header": Object {
      "contents": Array [
        Object {
          "color": "#00B172",
          "size": "lg",
          "text": "Provide more detail",
          "type": "text",
        },
      ],
      "layout": "vertical",
      "paddingBottom": "none",
      "type": "box",
    },
    "type": "bubble",
  },
  Object {
    "body": Object {
      "contents": Array [
        Object {
          "text": "You can turn on notifications if you want Cofacts to notify you when someone replies to this message.",
          "type": "text",
          "wrap": true,
        },
      ],
      "layout": "vertical",
      "type": "box",
    },
    "footer": Object {
      "contents": Array [
        Object {
          "action": Object {
            "label": "Go to settings",
            "type": "uri",
            "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=setting&utm_source=rumors-line-bot&utm_medium=reply-request",
          },
          "color": "#00B172",
          "style": "primary",
          "type": "button",
        },
      ],
      "layout": "vertical",
      "spacing": "sm",
      "type": "box",
    },
    "header": Object {
      "contents": Array [
        Object {
          "text": "üîî  Receive updates",
          "type": "text",
          "wrap": true,
        },
      ],
      "layout": "vertical",
      "paddingBottom": "none",
      "type": "box",
    },
    "type": "bubble",
  },
  Object {
    "body": Object {
      "contents": Array [
        Object {
          "text": "We all get by with a little help from our friends üåü Share your question to friends, someone might be able to help!",
          "type": "text",
          "wrap": true,
        },
      ],
      "layout": "vertical",
      "type": "box",
    },
    "footer": Object {
      "contents": Array [
        Object {
          "action": Object {
            "label": "Share on LINE",
            "type": "uri",
            "uri": "line://msg/text/?Please%20help%20me%20verify%20if%20this%20is%20true%3A%20https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id",
          },
          "color": "#ffb600",
          "style": "primary",
          "type": "button",
        },
        Object {
          "action": Object {
            "label": "Share on Facebook",
            "type": "uri",
            "uri": "https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&hashtag=%23ReportedToCofacts&href=https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id",
          },
          "color": "#ffb600",
          "style": "primary",
          "type": "button",
        },
      ],
      "layout": "vertical",
      "spacing": "sm",
      "type": "box",
    },
    "type": "bubble",
  },
]
`;

exports[`should create a UserArticleLink when creating a Article 1`] = `
Array [
  Object {
    "_id": "_id",
    "articleId": "new-article-id",
    "createdAt": 2020-01-01T00:00:00.000Z,
    "lastViewedAt": 2020-01-01T00:00:00.000Z,
    "userId": "user-id-0",
  },
]
`;

exports[`should submit article if user agrees to submit: has AI reply 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "foo",
        "text": "Some text forwarded by the user",
        "type": "text",
      },
    ],
    "sessionId": 1577923200000,
  },
  "replies": Array [
    Object {
      "altText": "The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!",
              "type": "text",
              "wrap": true,
            },
            Object {
              "action": Object {
                "label": "View reported message",
                "type": "uri",
                "uri": "https://dev.cofacts.tw/article/new-article-id",
              },
              "margin": "md",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "This article is still under verification, please refrain from believing it for now. \\\\nBelow is the preliminary analysis result by the bot, hoping to provide you with some insights.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "This article is still under verification, please refrain from believing it for now. \\\\nBelow is the preliminary analysis result by the bot, hoping to provide you with some insights.",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "createdAt": undefined,
      "sender": Object {
        "iconUrl": "/static/img/aireply.png?cachebust=20230405",
        "name": "AI analysis",
      },
      "text": "Hello from ChatGPT",
      "type": "text",
    },
    Object {
      "altText": "After reading the automatic analysis by the bot above, you can:",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "After reading the automatic analysis by the bot above, you can:",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Your submission is now recorded at https://dev.cofacts.tw/article/new-article-id",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "Is the AI analysis helpful?",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "paddingAll": "xl",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "üëç Yes",
                    "type": "uri",
                    "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=aiReplyFeedback&aiResponseId=undefined&vote=UPVOTE",
                  },
                  "color": "#00B172",
                  "style": "primary",
                  "type": "button",
                },
                Object {
                  "action": Object {
                    "label": "üòï No",
                    "type": "uri",
                    "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=aiReplyFeedback&aiResponseId=undefined&vote=DOWNVOTE",
                  },
                  "color": "#FB5959",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#00B172",
                  "size": "lg",
                  "text": "Provide feedback to AI analysis",
                  "type": "text",
                },
              ],
              "layout": "vertical",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "It would help fact checkers a lot if you provide more detail :)",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "Provide detail",
                    "type": "uri",
                    "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=comment&articleId=new-article-id",
                  },
                  "color": "#00B172",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#00B172",
                  "size": "lg",
                  "text": "Provide more detail",
                  "type": "text",
                },
              ],
              "layout": "vertical",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "We all get by with a little help from our friends üåü Share your question to friends, someone might be able to help!",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "Share on LINE",
                    "type": "uri",
                    "uri": "line://msg/text/?Please%20help%20me%20verify%20if%20this%20is%20true%3A%20https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
                Object {
                  "action": Object {
                    "label": "Share on Facebook",
                    "type": "uri",
                    "uri": "https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&hashtag=%23ReportedToCofacts&href=https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`should submit article if user agrees to submit: has no AI reply 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "foo",
        "text": "Some text forwarded by the user",
        "type": "text",
      },
    ],
    "sessionId": 1577923200000,
  },
  "replies": Array [
    Object {
      "altText": "The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!",
              "type": "text",
              "wrap": true,
            },
            Object {
              "action": Object {
                "label": "View reported message",
                "type": "uri",
                "uri": "https://dev.cofacts.tw/article/new-article-id-2",
              },
              "margin": "md",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "In the meantime, you can:",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "In the meantime, you can:",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Your submission is now recorded at https://dev.cofacts.tw/article/new-article-id-2",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "It would help fact checkers a lot if you provide more detail :)",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "Provide detail",
                    "type": "uri",
                    "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=comment&articleId=new-article-id-2",
                  },
                  "color": "#00B172",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#00B172",
                  "size": "lg",
                  "text": "Provide more detail",
                  "type": "text",
                },
              ],
              "layout": "vertical",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "We all get by with a little help from our friends üåü Share your question to friends, someone might be able to help!",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "Share on LINE",
                    "type": "uri",
                    "uri": "line://msg/text/?Please%20help%20me%20verify%20if%20this%20is%20true%3A%20https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id-2",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
                Object {
                  "action": Object {
                    "label": "Share on Facebook",
                    "type": "uri",
                    "uri": "https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&hashtag=%23ReportedToCofacts&href=https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id-2",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;

exports[`should submit image article if user agrees to submit 1`] = `
Object {
  "context": Object {
    "msgs": Array [
      Object {
        "id": "6530038889933",
        "type": "image",
      },
    ],
    "sessionId": 1577923200000,
  },
  "replies": Array [
    Object {
      "altText": "The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "The message has now been recorded at Cofacts for volunteers to fact-check. Thank you for submitting!",
              "type": "text",
              "wrap": true,
            },
            Object {
              "action": Object {
                "label": "View reported message",
                "type": "uri",
                "uri": "https://dev.cofacts.tw/article/new-article-id",
              },
              "margin": "md",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "In the meantime, you can:",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "In the meantime, you can:",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Your submission is now recorded at https://dev.cofacts.tw/article/new-article-id",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "It would help fact checkers a lot if you provide more detail :)",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "Provide detail",
                    "type": "uri",
                    "uri": "https://liff.line.me/1563196602-X6mLdDkW?p=comment&articleId=new-article-id",
                  },
                  "color": "#00B172",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#00B172",
                  "size": "lg",
                  "text": "Provide more detail",
                  "type": "text",
                },
              ],
              "layout": "vertical",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "text": "We all get by with a little help from our friends üåü Share your question to friends, someone might be able to help!",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "vertical",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "label": "Share on LINE",
                    "type": "uri",
                    "uri": "line://msg/text/?Please%20help%20me%20verify%20if%20this%20is%20true%3A%20https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
                Object {
                  "action": Object {
                    "label": "Share on Facebook",
                    "type": "uri",
                    "uri": "https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&hashtag=%23ReportedToCofacts&href=https%3A%2F%2Fdev.cofacts.tw%2Farticle%2Fnew-article-id",
                  },
                  "color": "#ffb600",
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "vertical",
              "spacing": "sm",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
}
`;
